improver/utilities/cube_extraction.py

```diff
@@ -250,3 +252,123 @@ def extract_subcube(cube, constraints, units=None, use_original_units=True):
+        )
+        cutout = cube.extract(x_constraint & y_constraint)
+
+    return cutout
+
+
+def subset_data(cube, grid_spec=None, site_list=None):
+    """Extract a spatial cutout or subset of sites from data
+    to generate suite reference outputs.
+
+    Args:
+        cube (iris.cube.Cube):
+            Input dataset
+        grid_spec (Dict[str, Dict[str, int]]):
+            Dictionary containing bounding grid points and an integer "thinning
+            factor" for each of UK and global grid, to create cutouts.  Eg a
+            "thinning factor" of 10 would mean every 10th point being taken for
+            the cutout.  The expected dictionary has keys that are spatial coordinate
+            names, with values that are dictionaries with "min", "max" and "thin" keys.
+        site_list (list):
+            List of WMO site IDs to extract.  These IDs must match the type and format
+            of the "wmo_id" coordinate on the input spot cube.
+
+    Returns:
+        iris.cube.Cube:
+            Subset of input cube as specified by input constraints
+
+    Raises:
+        ValueError:
+            If site_list is not provided for a spot data cube
+        ValueError:
+            If the spot data cube does not contain any of the required sites
+        ValueError:
+            If grid_spec is not provided for a gridded cube
+        ValueError:
+            If grid_spec does not contain entries for the spatial coordinates on
+            the input gridded data
+        ValueError:
+            If the grid_spec provided does not overlap with the cube domain
+    """
+    if cube.coords("spot_index"):
+        if site_list is None:
+            raise ValueError("site_list required to extract from spot data")
+
+        constraint = Constraint(coord_values={"wmo_id": lambda x: x in site_list})
+        result = cube.extract(constraint)
+        if result is None:
+            raise ValueError(
+                f"Cube does not contain any of the required sites: {site_list}"
+            )
+
+    else:
+        if grid_spec is None:
+            raise ValueError("grid_spec required to extract from gridded data")
+
+        x_coord = cube.coord(axis="x").name()
+        y_coord = cube.coord(axis="y").name()
+
+        for coord in [y_coord, x_coord]:
+            if coord not in grid_spec:
+                raise ValueError(
+                    f"Cube coordinates {y_coord}, {x_coord} are not present within "
+                    f"{grid_spec.keys()}"
+                )
+
+        cutout = _create_cutout(cube, grid_spec)
+
+        if cutout is None:
+            raise ValueError(
+                "Cube domain does not overlap with cutout specified:\n"
+                f"{x_coord}: {grid_spec[x_coord]}, {y_coord}: {grid_spec[y_coord]}"
+            )
+
+        thin_x = grid_spec[x_coord]["thin"]
+        thin_y = grid_spec[y_coord]["thin"]
+        result_list = CubeList()
+        for subcube in cutout.slices([y_coord, x_coord]):
```

Comment:
I've noticed that this can cause an issue, if the `cutout` returned only contains a single point on either the x or y coordinates. This is because this x or y coordinate is then not a dimension coordinate and this slicing results in the following error: `ValueError: Requested an iterator over a coordinate (longitude) which does not describe a dimension.`

Reply:
This is true, and it's a case I hadn't considered.

I actually don't think it's necessary to support this, but it would certainly be more useful to the user if the error raised was explicit about that (ie "single point outputs are not supported").  But the wording might get convoluted.  What do you think?

Reply:
OK, raising an error for this cause seems fine.